### PR TITLE
Client init functions in AK's Python SDK

### PR DESCRIPTION
**Pull request by daabr**
Users shouldn't copy-paste them across samples and projects.

### CODE REVIEW

File: internal/backend/oauth/oauth.go
Comment (tebeka) [inline internal/backend/oauth/oauth.go:1]: How is this file related to the PR?

File: runtimes/pythonrt/py-sdk/autokitteh.py [hunk 1]
```diff
@@ -1,11 +1,306 @@
...
+try:
+    from atlassian import Jira
+    from google.auth.transport.requests import Request
+    import google.oauth2.credentials as credentials
+    import google.oauth2.service_account as service_account
+    from googleapiclient.discovery import build
+    from jira import JIRA
+    import slack_sdk
+except ModuleNotFoundError:
+    pass
```
Comment (tebeka) [inline runtimes/pythonrt/py-sdk/autokitteh.py]: From the Python zen:
```
Errors should never pass silently.
Unless explicitly silenced.
```

Issue a warning if fails.

In general, I prefer to break this to a module with `__init__.py` and then `jira.py` and friends. Then each module will import what it needs.

File: runtimes/pythonrt/py-sdk/autokitteh.py [hunk 2]
```diff
@@ -1,11 +1,306 @@
...
+    Code samples:
+    https://github.com/atlassian-api/atlassian-python-api/tree/master/examples/jira
+
+    Args:
+        connection: AutoKitteh connection name.
+
+    Returns:
+        Atlassian-Python-API Jira client.
+    """
+    if not re.fullmatch(r"[A-Za-z_]\w*", connection):
```
Comment (tebeka) [inline runtimes/pythonrt/py-sdk/autokitteh.py]: Refactor to `is_valid_connection` function and use everywhere.

File: runtimes/pythonrt/py-sdk/autokitteh.py [hunk 3]
```diff
@@ -1,11 +1,306 @@
...
+            "token": {
+                "access_token": os.getenv(connection + "__oauth_AccessToken"),
+                "token_type": os.getenv(connection + "__oauth_TokenType"),
+            },
+        },
+        **kwargs,
+    )
+
+
+def pycontribs_jira_client(connection: str, **kwargs):
```
Comment (tebeka) [inline runtimes/pythonrt/py-sdk/autokitteh.py]: Why the `pycontribs_` prefix?

File: runtimes/pythonrt/py-sdk/autokitteh.py [hunk 4]
```diff
@@ -1,11 +1,306 @@
...
+            return Jira(url=base_url, token=secret, **kwargs)
+        return Jira(
+            url=base_url,
+            username=email,
+            password=secret,
+            cloud=True,
+            **kwargs,
+        )
+
+    raise RuntimeError(f'AutoKitteh connection "{connection}" not initialized')
```
Comment (tebeka) [inline runtimes/pythonrt/py-sdk/autokitteh.py]: Create our own exceptions or use an exception from the `jira` module (say https://jira.readthedocs.io/api.html#jira.exceptions.JIRAError)

File: runtimes/pythonrt/py-sdk/autokitteh.py [hunk 5]
```diff
@@ -1,11 +1,306 @@
...
+
+
+def _atlassian_jira_client_cloud_oauth2(connection: str, **kwargs):
+    """Initialize a Jira client for Atlassian Cloud using OAuth 2.0."""
+    expiry = os.getenv(connection + "__oauth_Expiry")
+    if not expiry:
+        raise RuntimeError(f'AutoKitteh connection "{connection}" not initialized')
+
+    iso8601 = re.sub(r"[ A-Z]+$", "", expiry)  # Convert from Go's time string.
```
Comment (tebeka) [inline runtimes/pythonrt/py-sdk/autokitteh.py]: Add a comment with example of before and after.
Or use `datetime.strptime`?

**Review by tebeka — CHANGES_REQUESTED** [review]
We need to discuss as a team the role of the `autokitteh` Python module and if your changes fit there. Other than that, looks good minus some comments.

If you need help converting `autokitteh.py` to a directory LMK.

File: internal/backend/oauth/oauth.go
Comment (daabr) [reply internal/backend/oauth/oauth.go:1]: Oops, removed

File: runtimes/pythonrt/py-sdk/autokitteh.py (hunk 1 above)
Comment (daabr) [reply runtimes/pythonrt/py-sdk/autokitteh.py]: > Unless explicitly silenced.

That's exactly what the try-except is for here. I improved the comment to clarify.

Breaking into sub-modules - yes, it makes sense. WIP.

File: runtimes/pythonrt/py-sdk/autokitteh.py (hunk 3 above)
Comment (daabr) [reply runtimes/pythonrt/py-sdk/autokitteh.py]: I thought it would be useful to distinguish between the 2 libs, but you're right, removing.

File: runtimes/pythonrt/py-sdk/autokitteh.py (hunk 5 above)
Comment (daabr) [reply runtimes/pythonrt/py-sdk/autokitteh.py]: I strongly prefer a standard function such as `fromisoformat` over a custom pattern in `strptime`. I though the var name `iso8601` would be enough, but I'll add a specific example in a comment instead.

File: runtimes/pythonrt/py-sdk/autokitteh.py (hunk 1 above)
Comment (daabr) [reply runtimes/pythonrt/py-sdk/autokitteh.py]: About the try-except - nm, I needed to add integration libs to the Python CI to avoid test failures.

File: internal/backend/oauth/oauth.go
Comment (tebeka) [inline internal/backend/oauth/oauth.go:1]: How are the changes here related to the PR?

File: runtimes/pythonrt/dev-requirements.txt [hunk 6]
```diff
@@ -2,3 +2,12 @@
 
 pytest ~= 8.0
 ruff ~= 0.2
+
+# AutoKitteh integrations
+
+atlassian-python-api ~= 3.41
```
Comment (tebeka) [inline runtimes/pythonrt/dev-requirements.txt:8]: Use `-r ../requirements.txt` ?

File: runtimes/pythonrt/py-sdk/autokitteh/__init__.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+"""AutoKitteh Python SDK."""
+
+from . import atlassian
+from .attr_dict import AttrDict
+from .decorators import activity
+from . import errors
+from . import google
```
Comment (tebeka) [inline runtimes/pythonrt/py-sdk/autokitteh/__init__.py]: Don't import `google` and `slack`. It'll force users to have the corresponding SDK installed.

**Review by tebeka — CHANGES_REQUESTED** [review]
See comments in `__init__.py`
Also need to change `py-sdk/pyproject.toml` to account for a directory:

```
[tool.setuptools.packages]
find={}
```

File: internal/backend/oauth/oauth.go
Comment (daabr) [reply internal/backend/oauth/oauth.go:1]: Probably a bad rebase, considering https://github.com/autokitteh/autokitteh/pull/398 was already merged into main? Fixed now

File: runtimes/pythonrt/py-sdk/autokitteh/__init__.py (hunk 7 above)
Comment (daabr) [reply runtimes/pythonrt/py-sdk/autokitteh/__init__.py]: done

File: runtimes/pythonrt/py-sdk/autokitteh/__init__.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+
+__all__ = [
+    "AttrDict",
+    "activity",
+    "atlassian",
+    "errors",
+    "google",
+    "slack",
+]
```
Comment (tebeka) [inline runtimes/pythonrt/py-sdk/autokitteh/__init__.py]:
```suggestion
__all__ = [
    "AttrDict",
    "activity",
    "errors",
]
```

File: runtimes/pythonrt/dev-requirements.txt [hunk 9]
```diff
@@ -2,3 +2,12 @@
...
+google-api-python-client ~= 2.122
+google-auth-httplib2 ~= 0.2
+google-auth-oauthlib ~= 1.2
+jira ~= 3.8
+slack-sdk ~= 3.27
```
Comment (tebeka) [inline runtimes/pythonrt/dev-requirements.txt:13]: These should be in `requirements.txt` since `ak` uses it to build the virtual environment.
Maybe you can update the comment at the top of `requirements.txt` to be clearer?

**Review by tebeka — APPROVED** [review]
Left two comments, please fix them. Pre-approving.

File: runtimes/pythonrt/py-sdk/autokitteh/__init__.py (hunk 8 above)
Comment (daabr) [reply runtimes/pythonrt/py-sdk/autokitteh/__init__.py]: done

File: runtimes/pythonrt/dev-requirements.txt (hunk 9 above)
Comment (daabr) [reply runtimes/pythonrt/dev-requirements.txt:13]: Removed entirely
